state-sync/aptos-data-client/src/aptosnet/mod.rs

```diff
@@ -146,46 +156,65 @@ impl AptosNetDataClient {
+        // Ensure we don't go over the maximum number of in-flight priority polls
+        if num_in_flight_polls >= self.data_client_config.max_num_in_flight_priority_polls {
+            return Ok(None);
         }
 
-        // Fetch all new peers (i.e., those not yet polled)
-        for peer in self.get_all_connected_peers()? {
-            if !self.peer_states.read().already_polled_peer(&peer) {
-                peers_to_poll.push(peer);
-            }
-        }
+        // Get all connected prioritized peers
+        let mut priority_peers = self.get_all_connected_peers()?;
+        priority_peers.retain(|peer| self.peer_states.read().is_priority_peer(peer));
 
-        // Handle regular peer polling
-        if peers_to_poll.is_empty() {
-            // Always try and poll at least one peer
-            if let Some(peer) = self.peer_states.write().oldest_polled_regular_peer() {
-                peers_to_poll.push(peer);
-            }
-        } else {
-            // Poll regular peers at a 1/3 reduced frequency
-            sample!(SampleRate::Frequency(3), {
-                if let Some(peer) = self.peer_states.write().oldest_polled_regular_peer() {
-                    peers_to_poll.push(peer);
-                }
-            });
+        self.select_peer_to_poll(priority_peers)
+    }
+
+    /// Fetches the next regular peer to poll. The peer will be marked
+    /// as having an in-flight request.
+    fn fetch_regular_peer_to_poll(&self) -> Result<Option<PeerNetworkId>, Error> {
+        // Fetch the number of in-flight polls
+        let num_in_flight_polls = self.peer_states.read().num_in_flight_regular_polls();
+        update_in_flight_metrics(REGULAR_PEER, num_in_flight_polls);
+
+        // Ensure we don't go over the maximum number of in-flight regular polls
+        if num_in_flight_polls >= self.data_client_config.max_num_in_flight_regular_polls {
+            return Ok(None);
         }
 
-        // Mark all peers as polled
-        for peer in &peers_to_poll {
-            self.peer_states.write().mark_peer_as_polled(peer);
+        // Get all connected regular peers
+        let mut regular_peers = self.get_all_connected_peers()?;
+        regular_peers.retain(|peer| !self.peer_states.read().is_priority_peer(peer));
+
+        self.select_peer_to_poll(regular_peers)
+    }
+
+    /// Randomly selects a peer to poll that does not have an in-flight request
+    fn select_peer_to_poll(
+        &self,
+        mut peers: Vec<PeerNetworkId>,
+    ) -> Result<Option<PeerNetworkId>, Error> {
+        // Identify the peers who do not already have in-flight requests.
+        peers.retain(|peer| !self.peer_states.read().existing_in_flight_request(peer));
+
+        // Select a peer at random for polling
+        let peer_to_poll = peers.choose(&mut rand::thread_rng());
+        if let Some(peer_to_poll) = peer_to_poll {
+            self.peer_states.write().new_in_flight_request(peer_to_poll);
```

Comment:
Any reason to mark this as inflight over here instead of `poll_peer` function? If feasible, let's move this to the `poll_peer` function, which semantically makes more sense as we mark this as `in_flight` before the polling begins and mark it as done afterwards.

Reply:
Done.